src/basic/proc-cmdline.c

```diff
@@ -10,14 +11,91 @@
 #include "initrd-util.h"
 #include "macro.h"
 #include "parse-util.h"
+#include "proc-cmdline-internal.h"
 #include "proc-cmdline.h"
 #include "process-util.h"
 #include "special.h"
 #include "string-util.h"
+#include "strv.h"
 #include "virt.h"
 
-int proc_cmdline(char **ret) {
+static void filter_pid1_args(
+                int argc,                /* input */
+                char **argv,             /* input, may be reordered by this function, should not be reused. */
+                const char **filtered) { /* optput */
+
+        int saved_optind, saved_opterr;
+        size_t n = 0;
+
+        assert(argc > 0);
+        assert(argv);
+        assert(filtered);
+
+        saved_optind = optind;
+        saved_opterr = opterr;
+
+        /* Resetting to 0 forces the invocation of an internal initialization routine of getopt_long()
+         * that checks for GNU extensions in optstring ('-' or '+' at the beginning). Here, we do not use
+         * the GNU extensions, but might be used previously. Hence, we need to always reset it. */
+        optind = 0;
+
+        /* Do not print an error message. */
+        opterr = 0;
+
+        /* Filter out all known options. */
+        while (getopt_long(argc, argv, SYSTEMD_GETOPT_SHORT_OPTIONS, systemd_getopt_options, NULL) >= 0)
+                ;
+
+        /* Also filter all strings start with '-'. */
+        for (int i = optind; i < argc; i++) {
+                if (argv[i][0] == '-')
+                        continue;
+
+                filtered[n++] = argv[i];
+        }
```

Comment:
I'm not sure how useful that part is. To actually filter an argument, we need to know if it takes an option or not. So by filtering unknown options, we may actually make the situation worse… What is the reasoning behind this filtering here?

Reply:
Yeah, but IIRC, in proc_cmdline_parse() or friends, we only accept options that are not prefixed with "-" or "--". So, I'd like to drop the options prefixed with "-" for safety. But, yeah, unexpected arguments prefixed with "-" are already filtered by the callback functions passed to `proc_cmdline_parse()`, hence the block may not be necessary. I will drop it.